### ISSUE TITLE
rc2 mode solver

### ISSUE THREAD

**Issue by joamatab** [flexcompute/tidy3d#1135]
For the mode solver in rc2, why is it assuming a 3D structure,

should I make it 2D or  use a Medium2D?

![image](https://github.com/flexcompute/tidy3d/assets/4514346/93c456e0-65e2-4adf-8e51-fe55256e4a86)

**Comment by momchil-flex** [comment]
I assume this is to be used in a 2D simulation. Tidy3D doesn't have a native separation between 2D and 3D. 2D is treated as a subclass of a 3D simulation, with a few small modifications to the backend. Because of this, there are some things that may not work perfectly (e.g. may depend on numerical precision) if you put a 2D structure in a 2D simulation (e.g. the discretization may go awry in some cases and it may try to subpixel it, or just not see the 2D structure but just the background medium). That is why we have started validating that all structures placed in the simulation are 3D to avoid such potential issues. The only time a structure can be 2D is if Medium2D is used, which is for things like graphene or a single PEC sheet, so not applicable in your case.

Basically what you need to do to avoid this warning is give some slab_thickness > 0. Ideally this should be larger than the grid step in the simulation.

**Comment by momchil-flex** [comment]
Ah ok I see the use case. Well, what do you want a slab with thickness of 0 to do? It should probably be completely ignored? If so, then in Tidy3D it is best to just not add this structure at all, i.e. have a check if thickness is 0 and avoid adding such structures if so. This is because if that structure of thickness zero happens to lie exactly on a grid point in the numerical grid, it will have an effect on that pixel and will not be completely ignored. Maybe it will be in some cases but I'm not 100% sure, it's kind of undefined behavior that could produce different results affected by numerical precision.

**Comment by joamatab** [comment]
a slab_thickness of 0 should be a strip waveguide
I think you mean normal direction to the waveguide, which you don't have an option to set to zero

our modesolver is basically based on your mode solver plugin, Lucas made some improvements on it when you released the plugin

@lucas-flexcompute

**Comment by momchil-flex** [comment]
Yeah, ok there may be something weird happening.

I also looked at Lucas' code and it does seem like slab_thickness = 0 should be already ignored. https://github.com/flexcompute/tidy3d/blob/677ef55ca21b6675f80c7404ea3b6ca9c33ee791/tidy3d/plugins/waveguide/rectangular_dielectric.py#L590-L600

**Comment by momchil-flex** [comment]
Ok so actually the warnings are created because of grid `override_structures` that the plugin makes. There are use cases in which these could be smaller than 3D indeed.

@caseyflex I would suggest we move this warning to `Simulation.structures`. So, allow users to create 2D structures without a warning, but warn if they try to add those to a simulation. That probably makes more sense going forward, if we have other cases in which 2D structures are actually ok (like they are in `Simulation.grid_spec.override_structures`). Also @tylerflex.

**Comment by tylerflex** [comment]
sounds good to me, alternatively we could automatically remove any override structures that are 2D?

**Comment by tylerflex** [comment]
Or do they affect the mesh by placing a cell boundary there?

**Comment by momchil-flex** [comment]
They do, and it's actually a good use case. For example I may want to force a cell boundary at a monitor location by just putting an override structure with the monitor geometry.
